config/load.go

```diff
@@ -0,0 +1,298 @@
+	// allow for nested options to be specified via environment variables
+	// e.g. pod.context = APPNAME_POD_CONTEXT
+	v := viper.NewWithOptions(viper.EnvKeyReplacer(strings.NewReplacer(".", "_", "-", "_")))
+
+	return load(cfg, v, cmd, configurations...)
+}
+
+func LoadAt(cfg Config, cmd *cobra.Command, path string, configuration any) error {
+	t := reflect.TypeOf(configuration)
+	config := reflect.StructOf([]reflect.StructField{{
+		Name: upperFirst(path),
+		Type: t,
+		Tag:  reflect.StructTag(fmt.Sprintf(`json:"%s" yaml:"%s" mapstructure:"%s"`, path, path, path)),
+	}})
+
+	value := reflect.New(config)
+	value.Elem().Field(0).Set(reflect.ValueOf(configuration))
+
+	return Load(cfg, cmd, value.Interface())
+}
+
+func upperFirst(p string) string {
+	if len(p) < 2 {
+		return strings.ToUpper(p)
+	}
+	return strings.ToUpper(p[0:1]) + p[1:]
+}
+
+func load(cfg Config, v *viper.Viper, cmd *cobra.Command, configurations ...any) error {
+	for _, cfg := range configurations {
+		if reflect.TypeOf(cfg).Kind() != reflect.Ptr {
+			return fmt.Errorf("LoadConfig cfg parameter must be a pointer, got: %s -- %v", reflect.TypeOf(cfg).Name(), cfg)
+		}
+	}
+
+	// priority order: viper.Set, flag, env, config, kv, defaults
+	// flags have already been loaded into viper by command construction
+
+	// check if user specified config; otherwise read all possible paths
+	if err := loadConfig(cfg, v, cfg.AppName, cfg.ConfigFile); err != nil {
+		var notFound *viper.ConfigFileNotFoundError
+		if errors.As(err, &notFound) {
+			cfg.Log.Debug("no config file found, using defaults")
+		} else {
+			return fmt.Errorf("unable to load config: %w", err)
+		}
+	}
+
+	// load environment variables
+	v.SetEnvPrefix(cfg.AppName)
+	v.AllowEmptyEnv(true)
+	v.AutomaticEnv()
+
+	appPrefix := cfg.AppName
+	if appPrefix != "" {
+		appPrefix += "."
+	}
+
+	flags := getFlagRefs(cmd)
+
+	for _, configuration := range configurations {
+		configureViper(cfg, v, reflect.ValueOf(configuration), flags, appPrefix, "")
+
+		// unmarshal fully populated viper object onto config
+		err := v.Unmarshal(configuration)
+		if err != nil {
+			return err
+		}
+
+		// Convert all populated config options to their internal application values ex: scope string => scopeOpt source.Scope
+		err = processPostConfig(configuration)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// configureViper loads the default configuration values into the viper instance, before the config values are read and parsed
+func configureViper(cfg Config, v *viper.Viper, value reflect.Value, flags flagRefs, appPrefix string, path string) {
+	if value.Type().Kind() == reflect.Ptr && value.Type().Elem().Kind() != reflect.Struct {
+		if flag, ok := flags[value.Pointer()]; ok {
+			cfg.Log.Trace(fmt.Sprintf("binding: %s = %v (flag)\n", strings.ToUpper(regexp.MustCompile("[^a-zA-Z0-9]").ReplaceAllString(appPrefix+path, "_")), value.Elem().Interface()))
```

Comment:
mind breaking this line out a bit? (there are a few others like this). It almost seems like this trace logger statement could get broken out into its own function.

Reply:
readability is all it would help. Even something simple like breaking `strings.ToUpper(regexp.MustCompile("[^a-zA-Z0-9]").ReplaceAllString(appPrefix+path, "_")` into another variable. I see that's done twice, so could be it's own small function.